### PR TITLE
Flatpak session (a.k.a. flatpak-based app fw) support.

### PR DESCRIPTION
**Pull request by klihub**
This patch set adds flatpak-session support to meta-flatpak on top of the existing basic flatpak support.

This mostly consists of

1) a daemon for monitoring configured flatpak remotes, pulling in/updating published apps
2) a systemd generator to set up 'flatpak sessions' for monitored flatpak repositories during boot
3) a helper binary to launch flatpak applications belonging to a 'flatpak session'
4) optionally a set of pre-declared flatpak remotes/repositories to preconfigure the images with (IOW set up remote URL, signing public key, and associated session user)

flatpak-session support has its dedicated DISTRO_FEATURE ('flatpak-session'). There is also a corresponding flatpak-session.inc configuration fragment which is used to enable flatpak-sessions.

By default flatpak-session support is disabled.

### CODE REVIEW

File: meta-flatpak/recipes-flatpak/flatpak-utils/files/flatpak-utils/src/flatpak.c [hunk 1]
```diff
@@ -0,0 +1,1011 @@
...
+        return def;
+
+    val = g_key_file_get_value(m, sec, key, NULL);
+
+    if (val == NULL)
+        return def;
+
+    i = strtol(val, &e, 10);
+
+    if (!e && !*e)
```
Comment (ipuustin) [inline meta-flatpak/recipes-flatpak/flatpak-utils/files/flatpak-utils/src/flatpak.c]: Should be `if (e && !*e)`?

Comment (klihub) [inline meta-flatpak/recipes-flatpak/flatpak-utils/files/flatpak-utils/src/flatpak.c]: Thanks, well spotted.

File: meta-flatpak/recipes-flatpak/flatpak-utils/flatpak-utils_git.bb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    ${systemd_unitdir}/system/flatpak-sessions.target \
+    ${systemd_unitdir}/system/flatpak-session@.service \
+    ${systemd_unitdir}/system/flatpak-update.service \
+"
+
+FILES_${PN}-dbg =+ "${base_libdir}/systemd/system-generators/.debug"
+
+SYSTEMD_PACKAGES      += "${PN}"
+SYSTEMD_SERVICE_${PN}  = "flatpak-sessions.target flatpak-update.service"
+SYSTEMD_AUTO_ENABLE    = "enable"
```
Comment (ipuustin) [inline meta-flatpak/recipes-flatpak/flatpak-utils/flatpak-utils_git.bb]: This is `enable` by default, no need to set it. And anyway if you set it, you might want to use `?=`.

File: meta-flatpak/recipes-flatpak/flatpak-utils/flatpak-utils_git.bb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+inherit autotools pkgconfig requires-systemd
+
+AUTO_LIBNAME_PKGS = ""
+
+S = "${WORKDIR}/git"
+
+EXTRA_OECONF += "--with-systemdunitdir=${systemd_unitdir}"
+
+# possible package configurations
+PACKAGECONFIG ??= ""
```
Comment (ipuustin) [inline meta-flatpak/recipes-flatpak/flatpak-utils/flatpak-utils_git.bb]: Hmm, why is this needed?

File: meta-flatpak/recipes-flatpak/flatpak-utils/flatpak-utils_0.0.bb [hunk 4]
```diff
@@ -7,16 +7,14 @@ LIC_FILES_CHKSUM = "file://LICENSE-BSD;md5=f9f435c1bd3a753365e799edf375fc42"
 DEPENDS = "flatpak systemd"
 
 SRC_URI = " \
-    git://git@github.com/klihub/flatpak-utils.git;protocol=https;branch=master \
+    file://flatpak-utils \
```
Comment (ipuustin) [inline meta-flatpak/recipes-flatpak/flatpak-utils/flatpak-utils_0.0.bb]: You shouldn't patch a file which you introduced in the same PR. Just change it.

File: meta-flatpak/recipes-flatpak/flatpak-utils/files/flatpak-utils/Makefile.am [hunk 5]
```diff
@@ -0,0 +1,32 @@
+SUBDIRS  = . src #orig
+doc_DATA = AUTHORS ChangeLog NEWS README
+
+# This is the only way with automake I know of to force 'check-git-hooks'
+# to be evaluated before 'all'. If there is a nicer way, I'm all ears...
+BUILT_SOURCES = install-git-hooks
```
Comment (ipuustin) [inline meta-flatpak/recipes-flatpak/flatpak-utils/files/flatpak-utils/Makefile.am]: You don't need to install git hooks, if you have the source distribution inside the layer. There is no corresponding `.git` directory.

File: meta-flatpak/recipes-flatpak/flatpak-utils/files/flatpak-utils/src/filesystem.c [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+#include <fcntl.h>
+#include <limits.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+
+#include "flatpak-session.h"
+
+
+int fsys_prepare_session(context_t *c)
```
Comment (ipuustin) [inline meta-flatpak/recipes-flatpak/flatpak-utils/files/flatpak-utils/src/filesystem.c]: You anyway need to link against glib (you are using `GError` etc. in the flatpak interfacing code). Have you checked if you could just just glib file system functions (see https://developer.gnome.org/glib/stable/glib-File-Utilities.html)? That would reduce the amount of custom code we have to maintain.

**Comment by ipuustin** [comment]
A README file for flatpak-utils would be nice. There are functions which are either dead code or used from somewhere I don't see, like `fpak_app_systemd_scope`. Also the restart logic is not fully clear. Is the whole daemon restarted when any flatpak app is updated? Or are you running one daemon for each app?

File: meta-flatpak/recipes-flatpak/flatpak-utils/flatpak-utils_git.bb (hunk 2 above)
Comment (klihub) [inline meta-flatpak/recipes-flatpak/flatpak-utils/flatpak-utils_git.bb]: Removed SYSTEMD_AUTO_ENABLE.

File: meta-flatpak/recipes-flatpak/flatpak-utils/flatpak-utils_git.bb (hunk 3 above)
Comment (klihub) [inline meta-flatpak/recipes-flatpak/flatpak-utils/flatpak-utils_git.bb]: Removed empty PACKAGECONFIG.

File: meta-flatpak/recipes-flatpak/flatpak-utils/flatpak-utils_0.0.bb (hunk 4 above)
Comment (klihub) [inline meta-flatpak/recipes-flatpak/flatpak-utils/flatpak-utils_0.0.bb]: Squashed commits.

File: meta-flatpak/recipes-flatpak/flatpak-utils/files/flatpak-utils/Makefile.am (hunk 5 above)
Comment (klihub) [inline meta-flatpak/recipes-flatpak/flatpak-utils/files/flatpak-utils/Makefile.am]: Ripped out githooks and removed associated rule.

File: meta-flatpak/recipes-flatpak/flatpak-utils/files/flatpak-utils/src/filesystem.c (hunk 6 above)
Comment (klihub) [inline meta-flatpak/recipes-flatpak/flatpak-utils/files/flatpak-utils/src/filesystem.c]: Sure I could have used. glib has a g-wrapper for every imaginable POSIX filesystem-related system call and then some. It was an arbitrary but deliberate choice to use glib where I have to, mostly for interfacing with libflatpak.

**Comment by klihub** [comment]
I've added a very brief README and removed the unused function. 

The restart logic is that whenever anything changes (gets updated) in a session, the whole session gets restarted. The actual restart is delegated to systemd. When a restart is desired, the daemon exits with a special exit code reserved for this purpose which systemd is let know about in the service file. Systemd will then restart the daemon which will then relaunch the applications within the session. At least that's the general idea...

**Comment by klihub** [comment]
Well, not quite yet. I need to still migrate some documentation over from the old repo. Also while I have added basic sanity tests, we still need to make sure they still work with latest meta-iotqa and CI infra. Once the tests work and the docs are in place, then I think this would be ready to go in.

**Comment by klihub** [comment]
Unfortunately running the tests for the flatpak-enabled gateway runtime image was not enabled at all. Fixed it. We'll have to check once the latest queued build is finished. Someone with supercow powers could show mercy on the currently running futile builds using the /stop REST API...

**Comment by klihub** [comment]
The test were run and have passed. I've also the added initial documentation snippets about flatpak support.

**Comment by klihub** [comment]
teat this please

**Comment by okartau** [comment]
test this please (there needs to be 's' in 1st word :-)

**Comment by klihub** [comment]
Thanks! Oh boy, I hate touchscreens...

**Comment by klihub** [comment]
@ipuustin Now, this could go in. If we don't want to enable flatpak sessions by default, we should leave [the config change commit ](https://github.com/klihub/intel-iot-refkit/commit/2c64cceff37a81deb14758d8369cda6e4fa8ea58) out.

**Comment by Arthil** [comment]
Retest this please.
